src/main/java/org/ow2/proactive/catalog/repository/specification/AbstractSpecification.java

```diff
@@ -75,9 +79,26 @@ protected void initCatalogObjectJoin(Root<CatalogObjectRevisionEntity> root, Cri
 
     protected void initMetadataJoin(Root<CatalogObjectRevisionEntity> root, CriteriaQuery<?> query,
             CriteriaBuilder cb) {
-        if (root.getJoins().size() == 1) {
-            metadataJoin = root.join("keyValueMetadataList", JoinType.INNER);
+        catalogObjectJoin = getOrCreateJoin(root, "catalogObject");
+        metadataJoin = getOrCreateJoin(root, "keyValueMetadataList");
+    }
+
+    protected void initBucketJoin(Root<CatalogObjectRevisionEntity> root, CriteriaQuery<?> query, CriteriaBuilder cb) {
+        catalogObjectJoin = getOrCreateJoin(root, "catalogObject");
+        bucketEntityJoin = getOrCreateJoin(catalogObjectJoin, "bucket");
+    }
+
+    private Join getOrCreateJoin(From from, String joinName) {
+        Optional<Join<CatalogObjectRevisionEntity, ?>> joinOptional = from.getJoins()
+                                                                          .stream()
+                                                                          .filter(join -> ((Join) join).getAttribute()
+                                                                                                       .getName()
+                                                                                                       .equals(joinName))
+                                                                          .findFirst();
```

Comment:
Just an idea, as I believe the first element that is returned here is quite important: shouldn't we ensure in this stream processing that we always have a deterministic and known behavior, such as sorting the stream before returning the first value found?